File: cpp/src/content_context.cpp
#include "content_context.hpp"
#include "convert.hpp"
#include <algorithm>
#include <iterator>

namespace longbridge {
namespace content {

using longbridge::convert::convert;

ContentContext::ContentContext()
  : ctx_(nullptr)
{
}

ContentContext::ContentContext(const lb_content_context_t* ctx)
{
  ctx_ = ctx;
  if (ctx_) {
    lb_content_context_retain(ctx_);
  }
}

ContentContext::ContentContext(const ContentContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_) {
    lb_content_context_retain(ctx_);
  }
}

ContentContext::ContentContext(ContentContext&& ctx)
{
  ctx_ = ctx.ctx_;
  ctx.ctx_ = nullptr;
}

ContentContext::~ContentContext()
{
  if (ctx_) {
    lb_content_context_release(ctx_);
  }
}

ContentContext&
ContentContext::operator=(const ContentContext& ctx)
{
  ctx_ = ctx.ctx_;
  if (ctx_) {
    lb_content_context_retain(ctx_);
  }
  return *this;
}

ContentContext
ContentContext::create(const Config& config)
{
  auto* ctx_ptr = lb_content_context_new(config);
  ContentContext ctx(ctx_ptr);
  if (ctx_ptr) {
    lb_content_context_release(ctx_ptr);
  }
  return ctx;
}

void
ContentContext::my_topics(
  const MyTopicsOptions& opts,
  AsyncCallback<ContentContext, std::vector<OwnedTopic>> callback) const
{
  const char* topic_type =
    opts.topic_type.empty() ? nullptr : opts.topic_type.c_str();
  lb_content_context_my_topics(
    ctx_,
    opts.page,
    opts.size,
    topic_type,
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<ContentContext, std::vector<OwnedTopic>>(
          res->userdata);
      ContentContext ctx((const lb_content_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto rows = (const lb_owned_topic_t*)res->data;
        std::vector<OwnedTopic> rows2;
        std::transform(rows,
                       rows + res->length,
                       std::back_inserter(rows2),
                       [](auto row) { return convert(&row); });

        (*callback_ptr)(
          AsyncResult<ContentContext, std::vector<OwnedTopic>>(
            ctx, std::move(status), &rows2));
      } else {
        (*callback_ptr)(
          AsyncResult<ContentContext, std::vector<OwnedTopic>>(
            ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<ContentContext, std::vector<OwnedTopic>>(callback));
}

void
ContentContext::create_topic(
  const CreateTopicOptions& opts,
  AsyncCallback<ContentContext, OwnedTopic> callback) const
{
  const char* topic_type =
    opts.topic_type.empty() ? nullptr : opts.topic_type.c_str();
  std::vector<const char*> tickers_cstr;
  for (const auto& t : opts.tickers) {
    tickers_cstr.push_back(t.c_str());
  }
  std::vector<const char*> hashtags_cstr;
  for (const auto& h : opts.hashtags) {
    hashtags_cstr.push_back(h.c_str());
  }
  lb_content_context_create_topic(
    ctx_,
    opts.title.c_str(),
    opts.body.c_str(),
    topic_type,
    tickers_cstr.empty() ? nullptr : tickers_cstr.data(),
    tickers_cstr.size(),
    hashtags_cstr.empty() ? nullptr : hashtags_cstr.data(),
    hashtags_cstr.size(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<ContentContext, OwnedTopic>(
          res->userdata);
      ContentContext ctx((const lb_content_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto resp = (const lb_owned_topic_t*)res->data;
        OwnedTopic result = convert(resp);

        (*callback_ptr)(
          AsyncResult<ContentContext, OwnedTopic>(
            ctx, std::move(status), &result));
      } else {
        (*callback_ptr)(
          AsyncResult<ContentContext, OwnedTopic>(
            ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<ContentContext, OwnedTopic>(callback));
}

void
ContentContext::topics(
  const std::string& symbol,
  AsyncCallback<ContentContext, std::vector<TopicItem>> callback) const
{
  lb_content_context_topics(
    ctx_,
    symbol.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<ContentContext, std::vector<TopicItem>>(
          res->userdata);
      ContentContext ctx((const lb_content_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto rows = (const lb_topic_item_t*)res->data;
        std::vector<TopicItem> rows2;
        std::transform(rows,
                       rows + res->length,
                       std::back_inserter(rows2),
                       [](auto row) { return convert(&row); });

        (*callback_ptr)(AsyncResult<ContentContext, std::vector<TopicItem>>(
          ctx, std::move(status), &rows2));
      } else {
        (*callback_ptr)(AsyncResult<ContentContext, std::vector<TopicItem>>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<ContentContext, std::vector<TopicItem>>(callback));
}

void
ContentContext::news(
  const std::string& symbol,
  AsyncCallback<ContentContext, std::vector<NewsItem>> callback) const
{
  lb_content_context_news(
    ctx_,
    symbol.c_str(),
    [](auto res) {
      auto callback_ptr =
        callback::get_async_callback<ContentContext, std::vector<NewsItem>>(
          res->userdata);
      ContentContext ctx((const lb_content_context_t*)res->ctx);
      Status status(res->error);

      if (status) {
        auto rows = (const lb_news_item_t*)res->data;
        std::vector<NewsItem> rows2;
        std::transform(rows,
                       rows + res->length,
                       std::back_inserter(rows2),
                       [](auto row) { return convert(&row); });

        (*callback_ptr)(AsyncResult<ContentContext, std::vector<NewsItem>>(
          ctx, std::move(status), &rows2));
      } else {
        (*callback_ptr)(AsyncResult<ContentContext, std::vector<NewsItem>>(
          ctx, std::move(status), nullptr));
      }
    },
    new AsyncCallback<ContentContext, std::vector<NewsItem>>(callback));
}

} // namespace content
} // namespace longbridge

File: java/src/content_context.rs
use std::sync::Arc;

use jni::{
    JNIEnv,
    objects::{JClass, JObject},
};
use longbridge::{
    Config,
    content::{ContentContext, CreateTopicOptions, ListMyTopicsOptions},
};

use crate::{
    async_util,
    error::jni_result,
    types::{FromJValue, JavaInteger, ObjectArray, get_field},
};

struct ContextObj {
    ctx: ContentContext,
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_newContentContext(
    mut env: JNIEnv,
    _class: JClass,
    config: i64,
) -> i64 {
    jni_result(&mut env, 0i64, |_env| {
        let config = Arc::new((*(config as *const Config)).clone());
        let ctx = ContentContext::new(config);
        Ok(Box::into_raw(Box::new(ContextObj { ctx })) as i64)
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_freeContentContext(
    _env: JNIEnv,
    _class: JClass,
    ctx: i64,
) {
    let _ = Box::from_raw(ctx as *mut ContextObj);
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_contentContextTopicsMine(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let page: Option<JavaInteger> = get_field(env, &opts, "page")?;
        let size: Option<JavaInteger> = get_field(env, &opts, "size")?;
        let topic_type: Option<String> = get_field(env, &opts, "topicType")?;
        async_util::execute(env, callback, async move {
            Ok(ObjectArray(
                context
                    .ctx
                    .topics_mine(ListMyTopicsOptions {
                        page: page.map(i32::from),
                        size: size.map(i32::from),
                        topic_type,
                    })
                    .await?,
            ))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_contentContextCreateTopic(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    opts: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let title: String = get_field(env, &opts, "title")?;
        let body: String = get_field(env, &opts, "body")?;
        let topic_type: Option<String> = get_field(env, &opts, "topicType")?;
        let tickers: Option<ObjectArray<String>> = get_field(env, &opts, "tickers")?;
        let hashtags: Option<ObjectArray<String>> = get_field(env, &opts, "hashtags")?;
        let license: Option<JavaInteger> = get_field(env, &opts, "license")?;
        async_util::execute(env, callback, async move {
            Ok(context
                .ctx
                .create_topic(CreateTopicOptions {
                    title,
                    body,
                    topic_type,
                    tickers: tickers.map(|a| a.0),
                    hashtags: hashtags.map(|a| a.0),
                    license: license.map(i32::from),
                })
                .await?)
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_contentContextTopics(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    symbol: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let symbol: String = FromJValue::from_jvalue(env, symbol.into())?;
        async_util::execute(env, callback, async move {
            Ok(ObjectArray(context.ctx.topics(symbol).await?))
        })?;
        Ok(())
    })
}

#[unsafe(no_mangle)]
pub unsafe extern "system" fn Java_com_longbridge_SdkNative_contentContextNews(
    mut env: JNIEnv,
    _class: JClass,
    context: i64,
    symbol: JObject,
    callback: JObject,
) {
    jni_result(&mut env, (), |env| {
        let context = &*(context as *const ContextObj);
        let symbol: String = FromJValue::from_jvalue(env, symbol.into())?;
        async_util::execute(env, callback, async move {
            Ok(ObjectArray(context.ctx.news(symbol).await?))
        })?;
        Ok(())
    })
}
